src/diffusers/pipelines/cogvideo/pipeline_cogvideox.py

```diff
@@ -409,6 +429,47 @@ def check_inputs(
                     f" {negative_prompt_embeds.shape}."
                 )
 
+    def fuse_qkv_projections(self) -> None:
+        r"""Enables fused QKV projections."""
+        self.fusing_transformer = True
+        self.transformer.fuse_qkv_projections()
+
+    def unfuse_qkv_projections(self) -> None:
+        r"""Disable QKV projection fusion if enabled."""
+        if not self.fusing_transformer:
+            logger.warning("The Transformer was not initially fused for QKV projections. Doing nothing.")
+        else:
+            self.transformer.unfuse_qkv_projections()
+            self.fusing_transformer = False
+
+    def _prepare_rotary_positional_embeddings(
+        self,
+        height: int,
+        width: int,
+        num_frames: int,
+        device: torch.device,
+        dtype: torch.dtype,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        grid_height = height // (self.vae_scale_factor_spatial * self.transformer.config.patch_size)
+        grid_width = width // (self.vae_scale_factor_spatial * self.transformer.config.patch_size)
+        base_size_width = 720 // (self.vae_scale_factor_spatial * self.transformer.config.patch_size)
+        base_size_height = 480 // (self.vae_scale_factor_spatial * self.transformer.config.patch_size)
+
+        grid_crops_coords = get_resize_crop_region_for_grid(
+            (grid_height, grid_width), base_size_width, base_size_height
+        )
+        freqs_cos, freqs_sin = get_3d_rotary_pos_embed(
+            embed_dim=self.transformer.config.attention_head_dim,
+            crops_coords=grid_crops_coords,
+            grid_size=(grid_height, grid_width),
+            temporal_size=num_frames,
+            use_real=True,
+        )
+
+        freqs_cos = freqs_cos.to(device=device, dtype=dtype)
```

Comment:
@a-r-r-o-w 

if we're going to undo the change about upcast in `apply_rotary_embedding` I think it is better to not downcast `freqs_cos` and `freqs_sin` here too, i.e. change to  `freqs_cos = freqs_cos.to(device=device)` here; otherwise I think the rotary embedding is created in float32, downcasted to float16 here and then upcast to float32 when applying - this cannot be good even if it does not have a noticeable impact 

can you test it out?

Reply:
Ah sorry, I overlooked this. Testing

Reply:
Also just noticed that `apply_rotary_emb` moves the embeddings to the correct device, so we can just remove these 2 lines entirely